Return nil for screen string index lookup instead of throwing error

Veratil opened the pull request:
There's often times when I'll have code to use a specific screen, and it always stops when the string lookup throws an error. This just makes it return nil instead so you can continue with the script (and handle nil return).

For example, in my rules I have a callback (which errors if the screen doesn't exist instead of putting it to the focused screen):
```lua
callback = function(c)
  c:move_to_tag(awful.tag.find_by_name(screen['DP-2'] or awful.screen.focused(), "EMAIL"))
end,
```

Also, I wasn't sure what value to return from this function, so I assumed 1.

actionless commented on objects/screen.c:
```diff
@@ -1040,7 +1040,8 @@ luaA_screen_module_index(lua_State *L)
             foreach(output, (*screen)->outputs)
                 if(A_STREQ(output->name, name))
                     return luaA_object_push(L, *screen);
-        luaL_error(L, "Unknown screen output name: %s", name);
```
mb still print the warning?

Veratil replied:
I could change it to `warn()`. Think that's okay?

psychon replied:
`luaA_warn` also adds a traceback and a timestamp (except on Lua 5.1 and LuaJIT)

Veratil replied:
Added warn. Tested with:

`$ awesome-client 'return screen["DP-2"]'`

Produces output (in my case):

```
Jun 21 08:44:07 zandaman /usr/libexec/gdm-x-session[16574]: 2018-06-21 08:44:07 [string "return screen["DP-2"]"]:1: W: Unknown screen output name: DP-2
Jun 21 08:44:07 zandaman /usr/libexec/gdm-x-session[16574]: stack traceback:
Jun 21 08:44:07 zandaman /usr/libexec/gdm-x-session[16574]: #011[C]: in function 'pcall'
Jun 21 08:44:07 zandaman /usr/libexec/gdm-x-session[16574]: #011/usr/local/share/awesome/lib/awful/remote.lua:26: in function </usr/local/share/awesome/lib/awful/remote.lua:20>
```

actionless commented:
@Veratil i think better just deprecation warning, without returning nil

Veratil commented:
> @Veratil i think better just deprecation warning, without returning nil

So deprecate and error out still? We could deprecate, warn, and push nil. :)

My main train of thought is of keeping screen similar to a normal table datatype (at least Lua 5.3, I'm not sure of previous behavior). If I do `t = {} return t[2]` then I get `nil`, not an error.

actionless commented:
i meant not to return nil if index is valid but just show deprecation warning what numeric screen lookup is deprecated

Veratil commented:
> i meant not to return nil if index is valid but just show deprecation warning what numeric screen lookup is deprecated

Oh I didn't mean return nil for everything. I just meant for invalid indexes we return nil instead of throwing an error.

actionless commented:
then it makes sense :+1:

Veratil commented:
Rebased on master and changed numeric indexing to return nil. I'm not sure what to return at that point, but I'm not entirely sure it's correct either.

actionless commented on objects/screen.c:
```diff
@@ -229,8 +229,13 @@ luaA_checkscreen(lua_State *L, int sidx)
-            luaL_error(L, "invalid screen number: %d (of %d existing)",
-                    screen, globalconf.screens.len);
+        {
+            //luaL_error(L, "invalid screen number: %d (of %d existing)",
```
and you need to remove the comments

blueyed commented:
Pushed to master with the comment removed (5d2d296).

blueyed commented:
Thank you!

psychon commented:
Sigh.

Could someone please fix `awful.widget.only_on_screen` and/or `tests/test-awful-widget-only_on_screen.lua`?

Output:

```
2018-07-13 18:28:38 lib/awful/widget/only_on_screen.lua:29: W: invalid screen number: 2 (of 1 existing)
stack traceback:
	[C]: in function 'pcall'
	lib/awful/widget/only_on_screen.lua:28: in function 'should_display_on'
	lib/awful/widget/only_on_screen.lua:42: in function <lib/awful/widget/only_on_screen.lua:41>
	[C]: in function 'xpcall'
	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>
	(...tail calls...)
	lib/gears/cache.lua:30: in function 'get'
	lib/wibox/widget/base.lua:295: in function 'fit_widget'
	lib/wibox/layout/align.lua:84: in function <lib/wibox/layout/align.lua:32>
	[C]: in function 'xpcall'
	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>
	(...tail calls...)
	lib/gears/cache.lua:30: in function <lib/gears/cache.lua:17>
	(...tail calls...)
	lib/wibox/hierarchy.lua:131: in function 'hierarchy_update'
	lib/wibox/hierarchy.lua:222: in function 'update'
	lib/wibox/drawable.lua:83: in function 'do_redraw'
	lib/wibox/drawable.lua:390: in function <lib/wibox/drawable.lua:388>
	[C]: in function 'xpcall'
	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>
	(...tail calls...)
	lib/gears/timer.lua:231: in function <lib/gears/timer.lua:229>
```
Code that causes this(`lib/awful/widget/only_on_screen.lua` line 28-30):
```lua
    local success, own_s = pcall(function()
        return capi.screen[self._private.screen]
    end)
```
If you want me to guess, something like the following beauty is needed to get rid of the warning:
```diff
diff --git a/lib/awful/widget/only_on_screen.lua b/lib/awful/widget/only_on_screen.lua
index 961f88699..6d8aae1d7 100644
--- a/lib/awful/widget/only_on_screen.lua
+++ b/lib/awful/widget/only_on_screen.lua
@@ -25,10 +25,12 @@ local function should_display_on(self, s)
     if not self._private.widget then
         return false
     end
-    local success, own_s = pcall(function()
-        return capi.screen[self._private.screen]
-    end)
-    return success and own_s == s
+    local own_s = self._private.screen
+    if type(own_s) == "number" and (own_s < 1 or own_s > capi.screen.count()) then
+        -- Invalid screen number
+        return false
+    end
+    return capi.screen[self._private.screen] == s
 end
 
 -- Layout this layout
```
However, this now runs into the following while generating API documentation, and now I'll let someone else clean up this mess:
```
2018-07-13 18:35:35 E: awesome: Error during a protected call: ...hon/projects/awesome/lib/awful/widget/only_on_screen.lua:28: attempt to call field 'count' (a number value)
stack traceback:
	...hon/projects/awesome/lib/awful/widget/only_on_screen.lua:28: in function 'should_display_on'
	...hon/projects/awesome/lib/awful/widget/only_on_screen.lua:44: in function <...hon/projects/awesome/lib/awful/widget/only_on_screen.lua:43>
	[C]: in function 'xpcall'
	/home/psychon/projects/awesome/lib/gears/protected_call.lua:36: in function </home/psychon/projects/awesome/lib/gears/protected_call.lua:35>
	(...tail calls...)
	/home/psychon/projects/awesome/lib/gears/cache.lua:30: in function 'get'
	/home/psychon/projects/awesome/lib/wibox/widget/base.lua:295: in function </home/psychon/projects/awesome/lib/wibox/widget/base.lua:282>
	(...tail calls...)
	[C]: in function 'xpcall'
	/home/psychon/projects/awesome/lib/gears/protected_call.lua:36: in function </home/psychon/projects/awesome/lib/gears/protected_call.lua:35>
	(...tail calls...)
	/home/psychon/projects/awesome/lib/gears/cache.lua:30: in function 'get'
	...
	[C]: in function 'xpcall'
	/home/psychon/projects/awesome/lib/gears/protected_call.lua:36: in function </home/psychon/projects/awesome/lib/gears/protected_call.lua:35>
	(...tail calls...)
	/home/psychon/projects/awesome/lib/gears/cache.lua:30: in function </home/psychon/projects/awesome/lib/gears/cache.lua:17>
	(...tail calls...)
	/home/psychon/projects/awesome/lib/wibox/hierarchy.lua:131: in function 'hierarchy_update'
	/home/psychon/projects/awesome/lib/wibox/hierarchy.lua:139: in function 'hierarchy_update'
	/home/psychon/projects/awesome/lib/wibox/hierarchy.lua:222: in function 'update'
	/home/psychon/projects/awesome/lib/wibox/hierarchy.lua:208: in function 'new'
	/home/psychon/projects/awesome/lib/wibox/widget/init.lua:43: in function 'draw_to_cairo_context'
	/home/psychon/projects/awesome/lib/wibox/widget/init.lua:56: in function 'draw_to_svg_file'
	...ome/tests/examples/wibox/container/defaults/template.lua:74: in main chunk
	[C]: in ?
```
It looks like Elv13's stub for `screen` is wrong in that `count` is a number instead of a function, but I cannot find the implementation....
Ah, found it. It's in `tests/examples/shims/screen.lua`:
```diff
diff --git a/tests/examples/shims/screen.lua b/tests/examples/shims/screen.lua
index 6e5d59dd8..f015dab05 100644
--- a/tests/examples/shims/screen.lua
+++ b/tests/examples/shims/screen.lua
@@ -2,7 +2,9 @@ local gears_obj = require("gears.object")
 
 local screen, meta = awesome._shim_fake_class()
 
-screen.count = 1
+function screen.count()
+    return 1
+end
 
 local function create_screen(args)
     local s = gears_obj()
```

actionless commented:
> I can live with the pcall() "behavior change". I searched github for users config affected and found none.
:-)
